src/services/branches/useBranch.tsx

```diff
@@ -0,0 +1,131 @@
+import { useQuery, UseQueryOptions } from '@tanstack/react-query'
+import { z } from 'zod'
+
+import {
+  RepoNotFoundErrorSchema,
+  RepoOwnerNotActivatedErrorSchema,
+} from 'services/repo'
+import Api from 'shared/api'
+import A from 'ui/A'
+
+export const BranchSchema = z
+  .object({
+    name: z.string(),
+    head: z
+      .object({
+        commitid: z.string().nullable(),
+      })
+      .nullable(),
+  })
+  .nullable()
+
+type BranchData = z.infer<typeof BranchSchema>
+
+const GetBranchSchema = z.object({
+  owner: z
+    .object({
+      repository: z
+        .discriminatedUnion('__typename', [
+          z.object({
+            __typename: z.literal('Repository'),
+            branch: BranchSchema,
+          }),
+          RepoNotFoundErrorSchema,
+          RepoOwnerNotActivatedErrorSchema,
+        ])
+        .nullable(),
+    })
+    .nullable(),
+})
+
+export interface UseBranchArgs {
+  provider: string
+  owner: string
+  repo: string
+  branch: string
+  opts?: UseQueryOptions<{ branch: BranchData }>
+}
+
+export const query = `
+query GetBranch($owner: String!, $repo: String!, $branch: String!) {
+  owner(username: $owner) {
+    repository(name: $repo) {
+      __typename
+      ... on Repository {
+        branch(name: $branch) {
+          name
+          head {
+            commitid
+          }
+        }
+      }
+      ... on NotFoundError {
+        message
+      }
+      ... on OwnerNotActivatedError {
+        message
+      }
+    }
+  }
+}`
+
+export const useBranch = ({
+  provider,
+  owner,
+  repo,
+  branch,
+  opts,
+}: UseBranchArgs) =>
+  useQuery({
+    queryKey: ['GetBranch', provider, owner, repo, branch, query],
+    queryFn: ({ signal }) =>
+      Api.graphql({
+        provider,
+        query,
+        signal,
+        variables: {
+          owner,
+          repo,
+          branch,
+        },
+      }).then((res) => {
```

Comment:
What would happen if we try to use this useBranch hook with a `select` statement somewhere in our codebase? If I recall correctly, the select would go first than the then, which is okay if we don't intend this hook with a `select`, is this the case?

Reply:
That is not how `select` works. So `queryFn` requires a promise to be returned if it's not a promise it will not work. `select` runs off the data which is returned by the `queryFn` so it would have to wait until the promise that is returned is resolved. Here's a little example showing off that `select` is not ran until the promise has resolved in the `queryFn`.

> [!NOTE]
> _This code follows the same conventions that occur when you use `Api.graphql`_

![Screenshot 2023-08-28 at 10 20 37 AM](https://github.com/codecov/gazebo/assets/105234307/fa7cd239-8960-4e1f-bdb0-4a090abf2154)

![Screenshot 2023-08-28 at 10 18 51 AM](https://github.com/codecov/gazebo/assets/105234307/1045c34e-32ae-43ad-bfc2-559c6270405d)